MULTIARCH-4515: add new endpoint to receive initrd.addrsize

AmadeusPodvratnik opened the pull request:
Add new parameter to return initrd.addrsize file overlayed by the current size of 
initrd.img and ignition.
This is only valid for s390x architecture and is necessary to support LPAR hypervisor (Classic and DPM).

This PR is the second required PR (First is https://github.com/openshift/assisted-image-service/pull/198).

This PR is for Assisted Installer, Agent-based Installer and HCP.
This new hypervisor support will be documented accordingly in the installation docs (https://issues.redhat.com/browse/MULTIARCH-4514)

## How was this code tested?
Running the initrd_test suite

## Assignees
@omertuc 

/cc @carbonin 
/cc 

## Links
[<!--
List any applicable links to related PRs or issues
-->](https://issues.redhat.com/browse/MULTIARCH-4515)[https://issues.redhat.com/browse/MULTIARCH-4515](https://issues.redhat.com/browse/MULTIARCH-4515)


## Checklist

- [x] Title and description added to both, commit and PR
- [x] Relevant issues have been associated
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit tests (note that code changes require unit tests)

AmadeusPodvratnik commented:
/ assign @omertuc

AmadeusPodvratnik commented on README.md:
```diff
@@ -124,6 +124,7 @@ Downloads the RHCOS initrd with the ignition for the specified image appended.
 
 - `version`: indicates the version of the RHCOS base image to use (must match an entry in `RHCOS_VERSIONS`)
 - `arch`: the base image cpu architecture (must match an entry in `RHCOS_VERSIONS`)
+- `addrsize`: indicates if the initrd.addrsize (`1`) file instead of initrd.img (`0`) should be returned. Only valid for s390x archtiecture
```
@paul-maidment Thank you for bringing this up. Maybe I bring some light into the thoughts from s390x community. Values like 0 and 1 are very common. In boot files like parm files values rd.neednet=1 or zfcp.allow_lun_scan=0 are used. Therefor I choose the value that way. But I understand that for users working with different architectures would get confused.

So what if we do it that way:
- change the parameter from 'addrsize' to 'file'
- value initrd.img or initrrd.addrsize with default initrd.img

Is this a feasible way to go?

omertuc replied:
It should just be a boolean. true or false

omertuc replied:
See below suggestion for my recommended phrasing. I mostly didn't like the use of "returned" since it's a too generic word and doesn't give much context of when and what

omertuc replied:
>from s390x community. Values like 0 and 1 are very common

This is the semantics of our HTTP service API, the conventions of the s390x community / kernel boot arguments don't apply

omertuc commented on internal/handlers/handlers_suite_test.go:
```diff
@@ -20,6 +20,8 @@ func TestHandlers(t *testing.T) {
 
 func createTestISO() string {
 	filesDir, err := os.MkdirTemp("", "isotest")
+	initrdAddrsize := []byte{
```
what do these bytes mean exactly?

AmadeusPodvratnik replied:
I choose the values for testing purpose to make it easier to debug and verify. If you checking the ISO having the initrd.addrsize it contains the values based on the kernel.img and initrd.img (the loading address of the initrd.img depends on the size of the kernel alligned to MiB boundary to get the PSW for the initrd).

omertuc replied:
Are they arbitrary though? if they have some sort of meaning/format, please use something like e.g.:

```go
bs := make([]byte, 8)
binary.LittleEndian.PutUint64(bs, uint64(64))
```

 To generate them

omertuc commented on internal/handlers/initrd.go:
```diff
@@ -35,6 +39,12 @@ func (h *initrdHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
+	addrsize := r.URL.Query().Get("addrsize")
+	if addrsize == "1" && arch != "s390x" {
+		httpErrorf(w, http.StatusBadRequest, "'addrsize' parameter only valid for s390x architecture")
```
```suggestion
		httpErrorf(w, http.StatusBadRequest, "'addrsize' parameter only valid for the s390x architecture")
```

omertuc commented on internal/handlers/initrd.go:
```diff
@@ -71,11 +81,49 @@ func (h *initrdHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	aReader := initrdReader
+
+	// if not s390x and addrsize=1 than regular processing
+	if addrsize != "" && addrsize == "1" {
```
```suggestion
	if addrsize == "1" {
```

omertuc commented on README.md:
```diff
@@ -124,6 +124,7 @@ Downloads the RHCOS initrd with the ignition for the specified image appended.
 
 - `version`: indicates the version of the RHCOS base image to use (must match an entry in `RHCOS_VERSIONS`)
 - `arch`: the base image cpu architecture (must match an entry in `RHCOS_VERSIONS`)
+- `addrsize`: indicates if the initrd.addrsize (`1`) file instead of initrd.img (`0`) should be returned. Only valid for s390x archtiecture
```
It should just be a boolean. true or false

omertuc commented on internal/handlers/initrd.go:
```diff
@@ -71,11 +81,49 @@ func (h *initrdHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	// if not s390x and addrsize=1 than regular processing
+	if addrsize != "" && addrsize == "1" {
+		// calculate the addrsize and return initrd.addrsize file
+		// get current addrsize file from iso
+		fileName = fmt.Sprintf("%s-initrd.addrsize", imageID)
+		addrsizeReader, err := isoeditor.GetFileFromISO(isoPath, initrdAddrsizePathInISO)
+		if err != nil {
+			log.Errorf("Error retrieving initrd.addsize file: %v, isoPath; %s\n", err, isoPath)
+			httpErrorf(w, http.StatusInternalServerError, "Failed to get initrd: %v", err)
+			return
+		}
+
+		// get the size of the initrd including the embedded ignition
+		sizeOfInitrd, err := initrdReader.Seek(0, io.SeekEnd)
+		if err != nil {
+			httpErrorf(w, http.StatusInternalServerError, "Failed to determine size of initrd: %v", err)
+			return
+		}
+		buf := new(bytes.Buffer)
+		binary.Write(buf, binary.BigEndian, sizeOfInitrd)
+
+		// create overlay reader for initrdsize
+		overlay_addrsize := overlay.Overlay{
+			Reader: bytes.NewReader(buf.Bytes()),
+			Offset: 8,
+			Length: 8,
+		}
+		aReader, err = overlay.NewOverlayReader(addrsizeReader, overlay_addrsize)
+		if err != nil {
+			httpErrorf(w, http.StatusInternalServerError, "Failed to create overlay reader for initrd.addrsize: %v", err)
+			return
+		}
+
+		defer addrsizeReader.Close()
```
Please move all the code my comment is replying to to its own separate, well-named function

omertuc commented on internal/handlers/initrd.go:
```diff
@@ -71,11 +81,49 @@ func (h *initrdHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		binary.Write(buf, binary.BigEndian, sizeOfInitrd)
+
+		// create overlay reader for initrdsize
+		overlay_addrsize := overlay.Overlay{
```
Variable name does not match standard golang style guidelines, please run linting

omertuc commented:
We've had a voice conversation to clear up some of my confusions about this PR, we've agreed to move this to a separate endpoint altogether, rather than having a query parameter change the file that's being returned

AmadeusPodvratnik commented:
@omertuc 
> We've had a voice conversation to clear up some of my confusions about this PR, we've agreed to move this to a separate endpoint altogether, rather than having a query parameter change the file that's being returned from initrd to addrsize

As discussed I added a new endpoint and changed the code accordingly. I will close all conversation and pls have a look at the new changes.

omertuc commented on internal/handlers/initrd_addrsize.go:
This file is almost entirely a copy of `internal/handlers/initrd.go`, could we move out their common parts into something like `initrd_utils.go` or something?

AmadeusPodvratnik replied:
I created an initrd_utils.go and put the common parts into it. I did not create an explicit test.go because the tests would be done via initrd-test.go and initrd_addrsize.go.

If you want to have a initrd_utils_test.go than I will move the tests form initrd_test into it. 
Thank you.

omertuc replied:
No need for a separate test file, thanks

omertuc commented on internal/handlers/images_test.go:
```diff
@@ -68,4 +68,31 @@ var _ = Describe("ServeHTTP", func() {
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(resp.StatusCode).To(Equal(http.StatusOK))
+		//respContent, err := io.ReadAll(resp.Body)
+		//Expect(err).NotTo(HaveOccurred())
```
Are these commented out intentionally? Please uncomment or remove entirely

omertuc commented on internal/handlers/initrd.go:
```diff
@@ -30,46 +27,14 @@ func (h *initrdHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		httpErrorf(w, code, err.Error())
 		return
 	}
+	// reader need to be closed
```
unnecessary comment

AmadeusPodvratnik replied:
Done - removed

omertuc commented on internal/handlers/initrd_addrsize.go:
```diff
@@ -0,0 +1,92 @@
+		httpErrorf(w, code, err.Error())
+		return
+	}
+	// reader need to be closed
```
remove unnecessary comment

AmadeusPodvratnik replied:
Done - removed

omertuc commented on internal/handlers/initrd_addrsize.go:
```diff
@@ -0,0 +1,92 @@
+	buf := new(bytes.Buffer)
+	binary.Write(buf, binary.BigEndian, sizeOfInitrd)
+
+	// create overlay reader for initrdsize
```
remove unnecessary comment

AmadeusPodvratnik replied:
Done.

omertuc commented on README.md:
```diff
@@ -131,6 +131,20 @@ Downloads the RHCOS initrd with the ignition for the specified image appended.
 
+### `GET /images/{image_id}/initrd-addrsize`
+
+Downloads the initrd.addrsize containing the psw of the initrd (8 bytes) and the size of the initrd (8 bytes). This if for s390x architecture only.
```
```suggestion
Downloads the initrd.addrsize containing the psw of the initrd (8 bytes) and the size of the initrd (8 bytes). This is for s390x architecture only.
```

AmadeusPodvratnik replied:
Done

omertuc commented on README.md:
```diff
@@ -131,6 +131,20 @@ Downloads the RHCOS initrd with the ignition for the specified image appended.
 
 - `Authorization`: this header is passed directly through to assisted service requests to handle RHSSO authentication
 
+### `GET /images/{image_id}/initrd-addrsize`
```
I would like to emphasize how specific this is, so let's have it be part of the name

```suggestion
### `GET /images/{image_id}/s390x-initrd-addrsize`
```

omertuc replied:
Or are there other IBM architectures which this might be relevant to in the future?

AmadeusPodvratnik replied:
No yet ... so for the time being s390x only.

AmadeusPodvratnik replied:
Done

omertuc commented on README.md:
```diff
@@ -131,6 +131,20 @@ Downloads the RHCOS initrd with the ignition for the specified image appended.
+#### Query parameters
+
+- `version`: indicates the version of the RHCOS base image to use (must match an entry in `RHCOS_VERSIONS`)
+- `arch`: the base image cpu architecture (must match an entry in `RHCOS_VERSIONS`)
```
```suggestion
- `arch`: the base image cpu architecture (for this endpoint, this must be set to s390x)
```

AmadeusPodvratnik replied:
Done

omertuc commented on internal/handlers/handlers_suite_test.go:
```diff
@@ -31,6 +33,7 @@ func createTestISO() string {
 	Expect(os.WriteFile(filepath.Join(filesDir, "images/pxeboot/rootfs.img"), []byte("this is rootfs"), 0600)).To(Succeed())
 	Expect(os.WriteFile(filepath.Join(filesDir, "images/pxeboot/vmlinuz"), []byte("this is kernel"), 0600)).To(Succeed())
 	Expect(os.WriteFile(filepath.Join(filesDir, "images/pxeboot/initrd.img"), []byte("this is initrd"), 0600)).To(Succeed())
+	Expect(os.WriteFile(filepath.Join(filesDir, "images/initrd.addrsize"), initrdAddrsize, 0600)).To(Succeed())
```
The variable is unnecessary

```suggestion
	Expect(os.WriteFile(filepath.Join(filesDir, "images/initrd.addrsize"), []byte{1, 2, 3, 4, 5, 6, 7, 8, 1, 2, 3, 4, 5, 6, 7, 8}, 0600)).To(Succeed())
```

AmadeusPodvratnik replied:
Done

omertuc commented on internal/handlers/images.go:
```diff
@@ -14,11 +14,12 @@ import (
+	byID           http.Handler
+	byToken        http.Handler
+	initrd         http.Handler
+	initrdAddrsize http.Handler
```
Also here, let's rename to `s390xInitrdAddrsize`, etc

AmadeusPodvratnik replied:
Done

omertuc commented on internal/handlers/images_test.go:
```diff
@@ -68,4 +68,31 @@ var _ = Describe("ServeHTTP", func() {
+
+	It("calls the initrdAddrsize handler ServeHTTP", func() {
+		imageID := "bf25292a-dddd-49dc-ab9c-3fb4c1f07071"
+		stubISOHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			Fail("ISO handler should not have been called")
+		})
+		stubInitrdAddrsize := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			Expect(r.Method).To(Equal(http.MethodGet))
+			Expect(r.URL.Path).To(Equal(fmt.Sprintf("/images/%s/initrd-addrsize", imageID)))
+			http.ServeContent(w, r, "initrd.addrsize", time.Now(), strings.NewReader("initrdcontent"))
+		})
+
+		imageHandler := &ImageHandler{
+			long:           stubISOHandler,
+			initrdAddrsize: stubInitrdAddrsize,
+		}
```
Would this work as well?

```suggestion
		stubInitrdAddrsize := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			Expect(r.Method).To(Equal(http.MethodGet))
			Expect(r.URL.Path).To(Equal(fmt.Sprintf("/images/%s/initrd-addrsize", imageID)))
			http.ServeContent(w, r, "initrd.addrsize", time.Now(), strings.NewReader("initrdcontent"))
		})

		imageHandler := &ImageHandler{
			initrdAddrsize: stubInitrdAddrsize,
		}
```

omertuc commented on internal/handlers/initrd_utils.go:
```diff
@@ -0,0 +1,61 @@
+)
+
+func initrdOverlayReader(imageStore imagestore.ImageStore, client *AssistedServiceClient, r *http.Request) (overlay.OverlayReader, string, int, error) {
+
```
```suggestion
```

AmadeusPodvratnik replied:
Done

omertuc commented on internal/handlers/initrd_addrsize.go:
```diff
@@ -0,0 +1,92 @@
+		return
+	}
+
+	buf := new(bytes.Buffer)
```
```suggestion
	addrsizeBytes := new(bytes.Buffer)
```

AmadeusPodvratnik replied:
Done

omertuc commented on internal/handlers/initrd_addrsize.go:
```diff
@@ -0,0 +1,92 @@
+	// create overlay reader for initrdsize
+	overlayAddrsize := overlay.Overlay{
+		Reader: bytes.NewReader(buf.Bytes()),
+		Offset: 8,
```
Why 8?

omertuc commented on internal/handlers/initrd_addrsize.go:
```diff
@@ -0,0 +1,92 @@
+	overlayAddrsize := overlay.Overlay{
+		Reader: bytes.NewReader(buf.Bytes()),
+		Offset: 8,
+		Length: 8,
```
```suggestion
		Length: len(addrsizeBytes),
```

omertuc replied:
Yep

omertuc commented on internal/handlers/initrd_addrsize.go:
```diff
@@ -0,0 +1,92 @@
+	binary.Write(buf, binary.BigEndian, sizeOfInitrd)
+
+	// create overlay reader for initrdsize
+	overlayAddrsize := overlay.Overlay{
+		Reader: bytes.NewReader(buf.Bytes()),
+		Offset: 8,
+		Length: 8,
+	}
+
+	aReader, err := overlay.NewOverlayReader(addrsizeReader, overlayAddrsize)
+	if err != nil {
+		httpErrorf(w, http.StatusInternalServerError, "Failed to create overlay reader for initrd.addrsize: %v", err)
+		return
+	}
+
+	modTime, err := http.ParseTime(lastModified)
+	if err != nil {
+		log.Warnf("Error parsing last modified time %s: %v", lastModified, err)
+		modTime = time.Now()
+	}
```
Do we really need this overlaay? Can't we just serve the addrsize bytes buffer?

AmadeusPodvratnik replied:
Yes ... because I need to take the initrd.addrsize from file and copy the size of initrd appended with the ignition into the last 8 byte and give return the new initrd.addrsize file back.

omertuc replied:
Ah I see. Isn't that just 8 bytes? Can't we just read them and append the size and serve the 16-byte buffer?

omertuc replied:
I'm not sure how that addresses my question. Why do a "live edit" of an existing file, when it's just 16 bytes? Read the first 8 bytes (the psw?), append the size of the new initrd, and serve that. The overlay is useful for editing huge files, it's over-engineered for 2 numbers

omertuc replied:
Why is this conversation marked as resolved?

AmadeusPodvratnik replied:
Sorry ... closed accidentally. Overlay removed and return new buffer.

omertuc commented:
/ok-to-test

omertuc commented:
/lgtm

omertuc commented:
Waiting for @carbonin 's approval

omertuc commented:
/lgtm

AmadeusPodvratnik commented:
/test edge-e2e-ai-operator-ztp

omertuc commented:
/retest

AmadeusPodvratnik commented on internal/handlers/initrd_addrsize.go:
```diff
@@ -0,0 +1,90 @@
+		modTime = time.Now()
+	}
+	w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=%s", fileName))
+	http.ServeContent(w, r, fileName, modTime, bytes.NewReader(append(initrdPSW, addrsizeBytes.Bytes()...)))
```
io.MultiReader returns an io.Reader but http.ServeContent requires an io.ReadSeeker and conversation will require several lines of code. So there should be no benefit.

AmadeusPodvratnik commented on internal/handlers/initrd_addrsize_test.go:
```diff
@@ -0,0 +1,139 @@
+		Expect(resp.StatusCode).To(Equal(http.StatusBadRequest))
+		respContent, err := io.ReadAll(resp.Body)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(respContent).To(Equal(unsupportedArch))
```
Remove this test case because endpoint is for s390x only.

AmadeusPodvratnik commented on internal/handlers/initrd_utils.go:
```diff
@@ -0,0 +1,60 @@
+	"github.com/openshift/assisted-image-service/pkg/overlay"
+)
+
+func initrdOverlayReader(imageStore imagestore.ImageStore, client *AssistedServiceClient, r *http.Request) (overlay.OverlayReader, string, int, error) {
```
Remove _utils.go and move to initrd.go.

AmadeusPodvratnik commented on internal/handlers/middleware.go:
```diff
@@ -33,8 +33,8 @@ func WithInitrdViaHTTP(handler http.Handler) http.HandlerFunc {
-			if !strings.HasSuffix(r.URL.Path, "/pxe-initrd") {
-				// Only "/pxe-initrd" is allowed to be fetched
+			if (!strings.HasSuffix(r.URL.Path, "/pxe-initrd")) && (!strings.HasSuffix(r.URL.Path, "/s390x-initrd-addrsize")) {
+				// Only "/pxe-initrd" and "/initrd-addrsize" is allowed to be fetched
```
Will remove for the time being because I'm not sure here.

AmadeusPodvratnik commented on internal/handlers/images_test.go:
```diff
@@ -68,4 +68,25 @@ var _ = Describe("ServeHTTP", func() {
+		resp, err := client.Get(fmt.Sprintf("%s/images/%s/s390x-initrd-addrsize", server.URL, imageID))
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(resp.StatusCode).To(Equal(http.StatusOK))
```
Content is checked.

AmadeusPodvratnik commented on internal/handlers/initrd_addrsize.go:
```diff
@@ -0,0 +1,90 @@
+	}
+	defer initrdReader.Close()
+
+	fmt.Println("get addrsize file:")
```
removed.

AmadeusPodvratnik commented on internal/handlers/initrd_addrsize.go:
```diff
@@ -0,0 +1,90 @@
+		httpErrorf(w, http.StatusBadRequest, "'version' parameter required for initrd download")
+		return
+	}
+	fmt.Println("Arch check:")
```
removed.

AmadeusPodvratnik commented on internal/handlers/images.go:
```diff
@@ -61,6 +62,12 @@ func NewImageHandler(is imagestore.ImageStore, assistedServiceClient *AssistedSe
 				client:     assistedServiceClient,
 			},
 		),
+		s390xInitrdAddrsize: stdmiddleware.Handler("/images/:imageID/initrd-addrsize", mdw,
```
Done

AmadeusPodvratnik commented on README.md:
```diff
@@ -131,6 +131,20 @@ Downloads the RHCOS initrd with the ignition for the specified image appended.
+#### Query parameters
+
+- `version`: indicates the version of the RHCOS base image to use (must match an entry in `RHCOS_VERSIONS`)
+- `arch`: the base image cpu architecture (for this endpoint, this must be set to s390x)
```
Done

carbonin commented:
@AmadeusPodvratnik @omertuc I fixed the rebase issue and squashed the commits with a more accurate message, can you both make sure everything looks right?

carbonin commented:
It looks like everything is, in fact, not okay.

@AmadeusPodvratnik did you perhaps forget to commit the change to add the `initrdOverlayReader` function back to the `initrd.go` file?
